### DISCUSSION TITLE
`vips_sink_screen` not available in bindings.

### DISCUSSION THREAD

**Discussion by vixalien** [libvips/libvips#3634] [Q&A]
Hi.

I wrote some bindings for libvips for GJS at https://github.com/vixalien/gvips.

I introspected all the operations from `Vips.Operation` by doing the following:

```js

  GObject
    .type_children(GObject.type_from_name("VipsOperation"))
    .forEach(add_docs);
```

but it seems the function `vips_sink_screen` is still not available, despite all others working. is it a private function? I saw it was utilised in vipsdisp(-tiny) and wanted to make a simple libvips display app in GTK using my bindings but the function seems unavailable. is there something I'm doing wrong?

**Comment by jcupitt** [comment]
Hi @vixalien, that sounds very cool!

I think the issue is that the `sink_screen` notify function comes from a background thread, so you can't do anything in that function without a bit of careful C work. For example, `vipsdisp-tiny` just adds the notify data to a queue of work for the main thread and returns immediately. You'd need to do something clever for Python too (thanks to the GIL). 

GTK will also only allow most work to be done on the main thread, so (I think!) you'll need to wrap `sink_screen` manually in C and include something like the `g_idle_add()` that vipsdisp uses.

**Reply by vixalien** [reply]
thanks for the quick answer. im not very proficient in C or libvips, so I'll ask a few more questions if you don't mind. I'm using javascript, not python. I think I'll try wrapping the function in C. do you have an example I can follow?

**Reply by vixalien** [reply]
also how is `vips_sink_screen` different from `vips_image_cache`

**Reply by jcupitt** [reply]
I would adapt the code in `vipsdisp-tiny`.

`image_cache` is synchronous, so when requests come in, the caller blocks until the pixels are ready.

`sink_screen` is asynchronous, so it always gives something back immediately, and some time later (perhaps a lot later!), a notification will arrive meaning "your pixels are finally ready".

If you play with `vipsdisp` or `nip2` a bit it should be clear: the screen repaints in the background and the UI doesn't need to wait for pixels to be calculated.

**Reply by vixalien** [reply]
yes. I saw that. I'm moving all the draw bits to C and calling them from javascript. I'll update this on how it goes
